### PR TITLE
policy: consistent enablement in agent and operator

### PR DESCRIPTION
**Pull request by dlapcevic**
Ref: https://github.com/cilium/cilium/issues/33360

The network policy enforcement system for K8s, Cilium and Cilium Clusterwide network policies can be disabled when not used, to reduce resource footprint of Cilium, and improve scalability and performance.

Conditions for policy disabled:
- EnablePolicy=never
- DisableCiliumEndpointCRD=true
- EnableK8sNetworkPolicy=false
- EnableCiliumNetworkPolicy=false
- EnableCiliumClusterwideNetworkPolicy=false
- IdentityAllocationMode=crd

**Important note**
When enabling, or re-enabling, network policies (going from disabled to enabled configuration), to avoid any traffic disruption, ensure to follow one of these two-step processes:
- First enable network policy configuration and update / restart the cilium-agent pods, before creating any network policies in the cluster.
- If network policies are already created in the cluster, avoid enabling all the configuration at once. First change DisableCiliumEndpointCRD to false and update / restart cilium-agent pods, before applying the rest of the configuration. Do not change EnablePolicy and EnableK8sNetworkPolicy, EnableCiliumNetworkPolicy or EnableCiliumNetworkPolicy before first changing DisableCiliumEndpointCRD to false and updating / restarting cilium-agent pods.

This is especially important for larger clusters that can take longer time to update cilium-agent daemonset and reconcile the system. The reason is that the network policy enforcement system on cilium-agent requires Cilium Endpoint CRD to enforce policies for pods on remote nodes. Therefore, if network policies and Cilium Endpoint CRD are enabled at the same time, some traffic between the nodes can be disrupted (dropped), because rolling update / restart of cilium-agent takes time and remote node Cilium Endpoint might not yet been created when on the local node the policy that requires it is already being enforced.

Signed-off-by: Dorde Lapcevic <[dordel@google.com](mailto:dordel@google.com)>

### CODE REVIEW

**Comment by joestringer** [comment]
Could you move the detailed explanation for safe upgrade into the docs? The release notes in the PR description are expected to be formatted as a single-line notification for users that they can then click to learn more from the PR. Consider that the text from the release note in the PR description ends up as just one bulletpoint in a list like the one here: https://github.com/cilium/cilium/releases/tag/v1.17.0-pre.2

**Comment by tamilmani1989** [comment]
Are these settings apply only for CRD based mode ? Can users enable network policy with etcd as backend and still keep `DisableCiliumEndpointCRD` to true?

File: pkg/option/features.go
Comment (tklauser) [inline pkg/option/features.go]: While at it, this could be simplified to use `cmp.Or`:

```go
func NetPolicySystemIsEnabled(cfg *DaemonConfig) bool {
	return cmp.Or(
		cfg.EnablePolicy != NeverEnforce,
		cfg.EnableK8sNetworkPolicy,
		cfg.EnableCiliumNetworkPolicy,
		cfg.EnableCiliumClusterwideNetworkPolicy,
		!cfg.DisableCiliumEndpointCRD,
		cfg.IdentityAllocationMode != IdentityAllocationModeCRD,
	)
```

File: operator/pkg/ciliumidentity/controller.go
Comment (bimmlerd) [inline operator/pkg/ciliumidentity/controller.go]: nit: could collapse this with the above and `cmp.Or`?

File: pkg/option/features.go
Comment (bimmlerd) [inline pkg/option/features.go]: nit: Maybe just `NetworkPolicyEnabled`?

File: operator/pkg/ciliumidentity/cell.go
Comment (bimmlerd) [inline operator/pkg/ciliumidentity/cell.go]: `EnableNetworkPolicy` or `EnableNetworkPolicyEnforcement` - the "system" seems redundant.

Also, shouldn't this be `DisableNetworkPolicy`? Default value for a `bool` in go is `false`, and I'm a bit afraid that it's too easy to "accidentially" disable this somehow

**Review by bimmlerd — CHANGES_REQUESTED** [review]
few nits

File: pkg/option/features.go
Comment (dlapcevic) [reply pkg/option/features.go]: Done.

File: operator/pkg/ciliumidentity/controller.go
Comment (dlapcevic) [reply operator/pkg/ciliumidentity/controller.go]: Done.

File: pkg/option/features.go
Comment (dlapcevic) [reply pkg/option/features.go]: Done.

File: operator/pkg/ciliumidentity/cell.go
Comment (dlapcevic) [reply operator/pkg/ciliumidentity/cell.go]: I agree. Changed to `DisableNetworkPolicy`.

**Comment by dlapcevic** [comment]
Response to comment (https://github.com/cilium/cilium/pull/36167#issuecomment-2499380929)

> EDIT: Actually on second thought I'm not sure we want to be documenting this for users.


Yes, it would be safer to do so. I updated the PR. We will not advertise this specific configuration to all the users, but rather those who come with the interest of scaling Cilium for other features, beyond the scale that network policy supports, can learn about it.

**Comment by dlapcevic** [comment]
Response to comment (https://github.com/cilium/cilium/pull/36167#issuecomment-2499557817)

> Are these settings apply only for CRD based mode ? Can users enable network policy with etcd as backend and still keep `DisableCiliumEndpointCRD` to true?

Currently, it would work only with CRD based mode, because the path is simpler for it, since IPCache is populated via the CRDs that are disabled based on existing configuration.

If there is interest in expanding this to etcd backend, we need to ensure that we have configuration that will in a similar way set up the system to get the same benefits (stop syncing IPCache for all pods to all nodes).

File: pkg/option/features.go
Comment (pippolo84) [inline pkg/option/features.go:13]: Since this is called from the operator, we should bind `EnablePolicy` and `EnableK8sNetworkPolicy` too to the operator `Viper` instance, as you already did for `EnableCiliumNetworkPolicy` and `EnableCiliumClusterwideNetworkPolicy` in https://github.com/cilium/cilium/pull/35049

**Review by pippolo84 — CHANGES_REQUESTED** [review]
Thanks! Left a comment about operator options binding.

File: pkg/option/features.go
Comment (dlapcevic) [reply pkg/option/features.go:13]: Done.

Comment (pippolo84) [reply pkg/option/features.go:13]: I see you've added `EnableK8sNetworkPolicy`, but shouldn't we bind `EnablePolicy` too?

Comment (dlapcevic) [reply pkg/option/features.go:13]: You're right. I confused it with the derived config.

It's there now.

**Review by pippolo84 — APPROVED** [review]
:rocket:

**Review by bimmlerd — APPROVED** [review]
code looks ok to me. I have no idea whether this is a direction the project wants to take, but I'll happily defer to a maintainer.

**Review by tklauser — APPROVED** [review]

**Comment by joestringer** [comment]
> I have no idea whether this is a direction the project wants to take, but I'll happily defer to a maintainer.

Just to chime in a bit here from one perspective: There is a developing set of use cases from some heavy users of Cilium who may choose to implement network policy in ways external to Cilium, so we're exploring how to enable those deployments to be more resource efficient. Right now I'm relying on folks like @dlapcevic and @tamilmani1989 to explore these options and help maintain Cilium in this configuration, and I consider it to be a power user configuration that we avoid explicitly documenting. I don't personally have a good picture of exactly how we will ensure that the core of Cilium continues to be maintainable with this mode, but if we don't try then we won't build that knowledge / expertise. However if folks _are_ using this mode, I'm expecting them to be heavily involved in development of this mode to actively improve Cilium with these configuration parameters. This is a bit of a cautious approach but it seems like a reasonable middle ground for now.

**Pull request merged**
